draft PR -- hourglass control (updated Lagrangian) for elasticity and plasticity

Shuaihao-Zhang opened the pull request:
Here, I will show the results when dealing with the hourglass control method for ULSPH.

Shuaihao-Zhang commented:
This is the method we discussed just now. We only calculate the inverse of shear strain for `particle i` when the return mapping is called, i.e., when `particle i` is plastic. Then, we can calculate the scale matrix. If `particle i `is elastic, the scale matrix is the identity matrix.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/4e093ef0-6b5b-4a02-bde4-00bc5f772776)

https://github.com/Xiangyu-Hu/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L545-L553

The result is as follows:
![2d_talor_bar](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/f64ce620-c3cc-469a-ab4b-d500a1794c03)

As Prof. Hu mentioned, if the particle is plastic, then the inverse of shear strain exists.

However, there are some other problems, and the simulation crash after several steps. And it looks like tensile instability.

I think this may because the scale matrix is too small at the bottom of the bar, as shown in the following figure. 
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/cbef22ec-3a33-4a60-9e06-4301e4b1ac02)
The color indicates the scale matrix. It's an identity matrix when the particle is elastic, and the magnitude is 1.4 in 2D. If the particle is plastic, the magnitude of the scale matrix is smaller than 1.4.

I suppose, if the scale matrix is very small, the effect of this hourglass control method is also very small. So hourglass and tensile instability occurs.

The figure below shows the plastic indicator.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/c18b205d-092a-40fc-bbe3-7f1be030ae6e)

Xiangyu-Hu commented:
Are you sure that you have 1/2 in front of (\phi_i + \phi_j)?

Xiangyu-Hu commented:
So the parameter \xi has not effect to the solution?

Xiangyu-Hu commented:
> This is the method we discussed just now. We only calculate the inverse of shear strain for `particle i` when the return mapping is called, i.e., when `particle i` is plastic. Then, we can calculate the scale matrix. If `particle i `is elastic, the scale matrix is the identity matrix. ![image](https://user-images.githubusercontent.com/43446342/285507113-4e093ef0-6b5b-4a02-bde4-00bc5f772776.png)
> 
> https://github.com/Xiangyu-Hu/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L545-L553
> 
> The result is as follows: ![2d_talor_bar](https://user-images.githubusercontent.com/43446342/285505343-f64ce620-c3cc-469a-ab4b-d500a1794c03.gif) [ ![2d_talor_bar](https://user-images.githubusercontent.com/43446342/285505343-f64ce620-c3cc-469a-ab4b-d500a1794c03.gif) ](https://user-images.githubusercontent.com/43446342/285505343-f64ce620-c3cc-469a-ab4b-d500a1794c03.gif) [ ](https://user-images.githubusercontent.com/43446342/285505343-f64ce620-c3cc-469a-ab4b-d500a1794c03.gif)
> 
> As Prof. Hu mentioned, if the particle is plastic, then the inverse of shear strain exists.
> 
> However, there are some other problems, and the simulation crash after several steps. And it looks like tensile instability.
> 
> I think this may because the scale matrix is too small at the bottom of the bar, as shown in the following figure. ![image](https://user-images.githubusercontent.com/43446342/285506293-cbef22ec-3a33-4a60-9e06-4301e4b1ac02.png) The color indicates the scale matrix. It's an identity matrix when the particle is elastic, and the magnitude is 1.4 in 2D. If the particle is plastic, the magnitude of the scale matrix is smaller than 1.4.
> 
> I suppose, if the scale matrix is very small, the effect of this hourglass control method is also very small. So hourglass and tensile instability occurs.
> 
> The figure below shows the plastic indicator. ![image](https://user-images.githubusercontent.com/43446342/285508841-c18b205d-092a-40fc-bbe3-7f1be030ae6e.png)

Also, in the first part of the equation should it be \tild{\sigma_i} + \tild{\sigma_j}?

Shuaihao-Zhang commented:
> Are you sure that you have 1/2 in front of (\phi_i + \phi_j)?

Yes, I do put 1/2 in the code. This actually doesn't matter because we have this coefficient in front of this integration.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/d17f8a5e-7f81-4f5f-b778-524a65bb696a)

Shuaihao-Zhang commented:
> So the parameter \xi has not effect to the solution?

The parameter `𝜉` do have effects. But the simulation still crash even I used a large `𝜉` . I think this is because the difference between the elastic correction and plastic correction is too large, due to the scale matrix `𝝋` for plastic particle is too small.

Shuaihao-Zhang commented:
> 
> Also, in the first part of the equation should it be \tild{\sigma_i} + \tild{\sigma_j}?

Yes, this should be the shear stress after return mapping. This is a typo here. The code is correct.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/b30a1a99-6915-4dd1-ab2c-6d4ad4f20cd4)

Xiangyu-Hu commented:
I have checked you code and found, besides quite intervened coding, a serious consistent issue.
That is. If I comment line 
https://github.com/Shuaihao-Zhang/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L533
to skip return mapping and change 
https://github.com/Shuaihao-Zhang/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L555
to 
Mat3d deviatoric_stress = J2_plasticity_.ReturnMappingShearStress(shear_stress_3D_[index_i]);

Then the scaling factor should be even we are applying 
https://github.com/Shuaihao-Zhang/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L548

However, the results still give crashed results.
Please have a check.

Xiangyu-Hu commented:
Also. in the TL version, we use full strain to obtain scaling factor. Please have a look.
https://github.com/Xiangyu-Hu/SPHinXsys/blob/bbe70dc9411ab91c98b5c08fa3d8848b65112bef/src/shared/particle_dynamics/solid_dynamics/inelastic_dynamics.cpp#L44-L47C49

Shuaihao-Zhang commented:
> I have checked you code and found, besides quite intervened coding, a serious consistent issue. That is. If I comment line https://github.com/Shuaihao-Zhang/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L533 to skip return mapping and change https://github.com/Shuaihao-Zhang/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L555 to Mat3d deviatoric_stress = J2_plasticity_.ReturnMappingShearStress(shear_stress_3D_[index_i]);
> 
> Then the scaling factor should be Identity even we are applying https://github.com/Shuaihao-Zhang/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L548
> 
> However, the simulation still gives crashed results. Please have a check.

(1) Sorry for the coding. I have deleted some unnecessary code blocks and comments to make it more clear.
(2) The plastic behavior is not only controlled by return mapping, but also by the constitutive model. If we want to transfer the material to elasticity, besides skip the return mapping, we also need to modify a line of code in the constitutive model.
https://github.com/Xiangyu-Hu/SPHinXsys/blob/2eac06a0ef5745f0c97d47e71832d9114a21e9ef/tests/user_examples/extra_src/shared/general_continuum/general_continuum.cpp#L92-L114
We just change the Line-112 to `return shear_stress_rate_elastic;`, then we can get this elastic behavior.
![2d_taylor_bar_elastic](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/8cdeb9af-0842-4cfb-b1cd-e882a943cf1c)
And the scale matrix is an identity matrix for all particles.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/71f27b1f-229e-478a-85ae-439c4baebec4)

Shuaihao-Zhang commented:
> Also. in the TL version, we use full strain to obtain scaling factor. Please have a look. https://github.com/Xiangyu-Hu/SPHinXsys/blob/bbe70dc9411ab91c98b5c08fa3d8848b65112bef/src/shared/particle_dynamics/solid_dynamics/inelastic_dynamics.cpp#L44-L47C49

I will check this in detail tomorrow.

DongWuTUM commented:
@Shuaihao-Zhang, the returning map you wrote is to return the deviatoric part. The inconsistency between the elastic and plastic behaviors may arise due to the incompressible assumption. Pls, check it. Could you write down the plastic constitutive relation formulation in detail?

DongWuTUM commented:
@Xiangyu-Hu @Xiangyu-Hu Form the derivation, we should use the shear strain to obtain the scaling factor. But we can have a try for full strain.

Xiangyu-Hu commented:
Further more, the TL version use the full strain tensor not only the shear tension for the scaling factor.
Please see 

> > I have checked you code and found, besides quite intervened coding, a serious consistent issue. That is. If I comment line https://github.com/Shuaihao-Zhang/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L533 to skip return mapping and change https://github.com/Shuaihao-Zhang/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L555 to Mat3d deviatoric_stress = J2_plasticity_.ReturnMappingShearStress(shear_stress_3D_[index_i]);
> > Then the scaling factor should be Identity even we are applying https://github.com/Shuaihao-Zhang/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L548
> > However, the simulation still gives crashed results. Please have a check.
> 
> (1) Sorry for the coding. I have deleted some unnecessary code blocks and comments to make it more clear. (2) The plastic behavior is not only controlled by return mapping, but also by the constitutive model. If we want to transfer the material to elasticity, besides skip the return mapping, we also need to modify a line of code in the constitutive model.
> 
> https://github.com/Xiangyu-Hu/SPHinXsys/blob/2eac06a0ef5745f0c97d47e71832d9114a21e9ef/tests/user_examples/extra_src/shared/general_continuum/general_continuum.cpp#L92-L114
> 
> We just change the Line-112 to `return shear_stress_rate_elastic;`, then we can get this elastic behavior.
> ![2d_taylor_bar_elastic](https://private-user-images.githubusercontent.com/43446342/285666241-8cdeb9af-0842-4cfb-b1cd-e882a943cf1c.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEwMTgxMDQsIm5iZiI6MTcwMTAxNzgwNCwicGF0aCI6Ii80MzQ0NjM0Mi8yODU2NjYyNDEtOGNkZWI5YWYtMDg0Mi00Y2ZiLWIxY2QtZTg4MmE5NDNjZjFjLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI2VDE2NTY0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljN2E1YzYxNmVmYmI1MzNlMDE3MTU3OGM0MDVlMjRiN2Y2ODBmZjQxYWU2Y2FjMmRlYjIxYWU4ZmFhOWQzMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GSUWv98lcktb-vgKFMndtvD9lUbSsOPi2qwsuLKANgQ)
> And the scale matrix is an identity matrix for all particles.
> ![image](https://private-user-images.githubusercontent.com/43446342/285666272-71f27b1f-229e-478a-85ae-439c4baebec4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEwMTgxMDQsIm5iZiI6MTcwMTAxNzgwNCwicGF0aCI6Ii80MzQ0NjM0Mi8yODU2NjYyNzItNzFmMjdiMWYtMjI5ZS00NzhhLTg1YWUtNDM5YzRiYWViZWM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI2VDE2NTY0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI5N2M2ZjUyNjA4ZTJjMjYyYzY1MmY4MTgyNzYyZTBiZTMzODhjNjYzN2IzZTM3NzdhOTk1ZjVjNDlmZTFjZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mKDG-fVbCjjmN73e6qBx0Aa_qSj0qiJIL1e05KAxSW4)

Sorry. If I do as you said the 

> > I have checked you code and found, besides quite intervened coding, a serious consistent issue. That is. If I comment line https://github.com/Shuaihao-Zhang/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L533 to skip return mapping and change https://github.com/Shuaihao-Zhang/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L555 to Mat3d deviatoric_stress = J2_plasticity_.ReturnMappingShearStress(shear_stress_3D_[index_i]);
> > Then the scaling factor should be Identity even we are applying https://github.com/Shuaihao-Zhang/SPHinXsys/blob/06d36e86df5301e93e3381d6e9ed49e9a845ff88/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L548
> > However, the simulation still gives crashed results. Please have a check.
> 
> (1) Sorry for the coding. I have deleted some unnecessary code blocks and comments to make it more clear. (2) The plastic behavior is not only controlled by return mapping, but also by the constitutive model. If we want to transfer the material to elasticity, besides skip the return mapping, we also need to modify a line of code in the constitutive model.
> 
> https://github.com/Xiangyu-Hu/SPHinXsys/blob/2eac06a0ef5745f0c97d47e71832d9114a21e9ef/tests/user_examples/extra_src/shared/general_continuum/general_continuum.cpp#L92-L114
> 
> We just change the Line-112 to `return shear_stress_rate_elastic;`, then we can get this elastic behavior.
> ![2d_taylor_bar_elastic](https://private-user-images.githubusercontent.com/43446342/285666241-8cdeb9af-0842-4cfb-b1cd-e882a943cf1c.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEwMTgxMDQsIm5iZiI6MTcwMTAxNzgwNCwicGF0aCI6Ii80MzQ0NjM0Mi8yODU2NjYyNDEtOGNkZWI5YWYtMDg0Mi00Y2ZiLWIxY2QtZTg4MmE5NDNjZjFjLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI2VDE2NTY0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljN2E1YzYxNmVmYmI1MzNlMDE3MTU3OGM0MDVlMjRiN2Y2ODBmZjQxYWU2Y2FjMmRlYjIxYWU4ZmFhOWQzMDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GSUWv98lcktb-vgKFMndtvD9lUbSsOPi2qwsuLKANgQ)
> And the scale matrix is an identity matrix for all particles.
> ![image](https://private-user-images.githubusercontent.com/43446342/285666272-71f27b1f-229e-478a-85ae-439c4baebec4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEwMTgxMDQsIm5iZiI6MTcwMTAxNzgwNCwicGF0aCI6Ii80MzQ0NjM0Mi8yODU2NjYyNzItNzFmMjdiMWYtMjI5ZS00NzhhLTg1YWUtNDM5YzRiYWViZWM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI2VDE2NTY0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI5N2M2ZjUyNjA4ZTJjMjYyYzY1MmY4MTgyNzYyZTBiZTMzODhjNjYzN2IzZTM3NzdhOTk1ZjVjNDlmZTFjZWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mKDG-fVbCjjmN73e6qBx0Aa_qSj0qiJIL1e05KAxSW4)

If I do as you said by:    return shear_stress_rate_elastic;
Then I do not comments line 533 but move it before, Mat3d deviatoric_stress = shear_stress_3D_[index_i];
The simulation now should not blowup. but it does. So, there seems still has consistent issue.

Xiangyu-Hu commented:
Because it does not blowup if we simple constrain scale factor to identity in any case.

Shuaihao-Zhang commented:
> If I do as you said by: return shear_stress_rate_elastic; Then I do not comments line 533 but move it before, Mat3d deviatoric_stress = shear_stress_3D_[index_i]; The simulation now should not blowup. but it does. So, there seems still has consistent issue.

To transfer to elasticity, we should switch to `return shear_stress_rate_elastic;` and also comment Line 533 `shear_stress_3D_[index_i] = J2_plasticity_.ReturnMappingShearStress(shear_stress_3D_[index_i]);` to skip return mapping.  If we don't, then the return mapping still works, which makes the stress-strain behavior is non-linear. The stress and strain after return mapping is very small, so the scale matrix `𝝋` will be small. Then effect the hourglass control is small, and the simulation crashes because (I guess) tensile instability, as I mention above.

Shuaihao-Zhang commented:
> @Shuaihao-Zhang, the returning map you wrote is to return the deviatoric part. The inconsistency between the elastic and plastic behaviors may arise due to the incompressible assumption. Pls, check it. Could you write down the plastic constitutive relation formulation in detail?

Yes, this is the constitutive for J2 plasticity.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/d22fd18f-8218-4f8e-9263-d2816761a195)
For J2 plasticity, the plastic part only related to the shear stress and the pressure part has nothing to do with the plasticity. This is why the return mapping only applied to the deviatoric part.

Shuaihao-Zhang commented:
> @Xiangyu-Hu @Shuaihao-Zhang Form the derivation, we should use the shear strain to obtain the scaling factor. But we can have a try for full strain.

@Xiangyu-Hu @DongWuTUM 
I have check the full strain, and there is a problem.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/2fd0db63-a4c7-4291-a72d-e7b53b369d0e)
For shear strain, we can get the shear strain form shear stress by dividing by `2G`. While for total strain, we can not simply get the total strain from total stress.

Xiangyu-Hu commented:
> > If I do as you said by: return shear_stress_rate_elastic; Then I do not comments line 533 but move it before, Mat3d deviatoric_stress = shear_stress_3D_[index_i]; The simulation now should not blowup. but it does. So, there seems still has consistent issue.
> 
> To transfer to elasticity, we should switch to `return shear_stress_rate_elastic;` and also comment Line 533 `shear_stress_3D_[index_i] = J2_plasticity_.ReturnMappingShearStress(shear_stress_3D_[index_i]);` to skip return mapping. If we don't, then the return mapping still works, which makes the stress-strain behavior is non-linear. The stress and strain after return mapping is very small, so the scale matrix `𝝋` will be small. Then effect the hourglass control is small, and the simulation crashes because (I guess) tensile instability, as I mention above.

 I think that I am wrong here because the shear stress became plastic at next time step then the scale factor is not identity anymore. 
However, the scaling in TL version should have check.

Xiangyu-Hu commented:
> > @Xiangyu-Hu @Shuaihao-Zhang Form the derivation, we should use the shear strain to obtain the scaling factor. But we can have a try for full strain.
> 
> @Xiangyu-Hu @DongWuTUM I have check the full strain, and there is a problem. ![image](https://private-user-images.githubusercontent.com/43446342/285700535-2fd0db63-a4c7-4291-a72d-e7b53b369d0e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEwNjM5NjYsIm5iZiI6MTcwMTA2MzY2NiwicGF0aCI6Ii80MzQ0NjM0Mi8yODU3MDA1MzUtMmZkMGRiNjMtYTRjNy00MjkxLWE3MmQtZTdiNTNiMzY5ZDBlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI3VDA1NDEwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3MThlMTJjMGNmOWNiNzlmY2U0ZmVlZmUzYjdhODliNWUyMzBiMjUyMjI3YzBjNzE0MzBmMmY1ODRiY2U1NGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.s90lQK_2CR1lxD9ZZQy-IeRuvIWNNEzEXiWXa1qt8vg) For shear strain, we can get the shear strain form shear stress by dividing by `2G`. While for total strain, we can not simply get the total strain from total stress.

You can have check the TL version. For the scaling factor, the method add the isotropic part back to the right hand cauchy strain.

Shuaihao-Zhang commented:
> It is clear that if we included the isotropic part the variation of the scaling factor will be smaller.

Sure, I will check this.

Xiangyu-Hu commented:
Another thing is that the transformation between 2 and 3 d may introduce consistent issue.

Xiangyu-Hu commented:
We better begin with the simplest plastic case first.

Shuaihao-Zhang commented:
> Another thing is that the transformation between 2 and 3 d may introduce consistent issue.

I think this is ok. We can use the case `zsh_2d_oscillating_beam_plastic` to test. The J2 plasticity material is used in this case. 
If we skip the return mapping and also switch to `return shear_stress_rate_elastic;` in the constitutive relative. We can get exactly the same result as the case `test_2d_oscillating_beam_ul`, which use the elastic material.

Xiangyu-Hu commented:
> > We better begin with the simplest plastic case first.
> 
> Sure, I will try to find some other cases, maybe one dimensional stretching or compression of metals.

No. As our issue is stability, we need just simply the material properties for testing.

Shuaihao-Zhang commented:
> > > We better begin with the simplest plastic case first.
> > 
> > 
> > Sure, I will try to find some other cases, maybe one dimensional stretching or compression of metals.
> 
> No. As our issue is stability, we need just simply the material properties for testing.

I guess J2 plasticity is one of the simplest plastic models. I will check.

DongWuTUM commented:
> > > If I do as you said by: return shear_stress_rate_elastic; Then I do not comments line 533 but move it before, Mat3d deviatoric_stress = shear_stress_3D_[index_i]; The simulation now should not blowup. but it does. So, there seems still has consistent issue.
> > 
> > 
> > To transfer to elasticity, we should switch to `return shear_stress_rate_elastic;` and also comment Line 533 `shear_stress_3D_[index_i] = J2_plasticity_.ReturnMappingShearStress(shear_stress_3D_[index_i]);` to skip return mapping. If we don't, then the return mapping still works, which makes the stress-strain behavior is non-linear. The stress and strain after return mapping is very small, so the scale matrix `𝝋` will be small. Then effect the hourglass control is small, and the simulation crashes because (I guess) tensile instability, as I mention above.
> 
> I think that I am wrong here because the shear stress became plastic at next time step then the scale factor is not identity anymore. However, the scaling in TL version should have check.

I think it is Okay that the scale factor is identity because we have checked whether it is the plastic stage or not just before calculating the scale factor. And it is just related to this time step, not the next step.

DongWuTUM commented:
> > > > We better begin with the simplest plastic case first.
> > > 
> > > 
> > > Sure, I will try to find some other cases, maybe one dimensional stretching or compression of metals.
> > 
> > 
> > No. As our issue is stability, we need just simply the material properties for testing.
> 
> I guess J2 plasticity is one of the simplest plastic models. I will check.

Yes, it is already the simplest.

DongWuTUM commented:
> > If I do as you said by: return shear_stress_rate_elastic; Then I do not comments line 533 but move it before, Mat3d deviatoric_stress = shear_stress_3D_[index_i]; The simulation now should not blowup. but it does. So, there seems still has consistent issue.
> 
> To transfer to elasticity, we should switch to `return shear_stress_rate_elastic;` and also comment Line 533 `shear_stress_3D_[index_i] = J2_plasticity_.ReturnMappingShearStress(shear_stress_3D_[index_i]);` to skip return mapping. If we don't, then the return mapping still works, which makes the stress-strain behavior is non-linear. The stress and strain after return mapping is very small, so the scale matrix `𝝋` will be small. Then effect the hourglass control is small, and the simulation crashes because (I guess) tensile instability, as I mention above.

I am still confused about why "the stress and strain after return mapping are very small, so the scale matrix `𝝋` will be small". If it is perfect plasticity, the shear strain after the return map should be the same as the elastic strain. Does it mean the elastic strain is very small? Please check this.

Shuaihao-Zhang commented:
> > > If I do as you said by: return shear_stress_rate_elastic; Then I do not comments line 533 but move it before, Mat3d deviatoric_stress = shear_stress_3D_[index_i]; The simulation now should not blowup. but it does. So, there seems still has consistent issue.
> > 
> > 
> > To transfer to elasticity, we should switch to `return shear_stress_rate_elastic;` and also comment Line 533 `shear_stress_3D_[index_i] = J2_plasticity_.ReturnMappingShearStress(shear_stress_3D_[index_i]);` to skip return mapping. If we don't, then the return mapping still works, which makes the stress-strain behavior is non-linear. The stress and strain after return mapping is very small, so the scale matrix `𝝋` will be small. Then effect the hourglass control is small, and the simulation crashes because (I guess) tensile instability, as I mention above.
> 
> I am still confused about why "the stress and strain after return mapping are very small, so the scale matrix `𝝋` will be small". If it is perfect plasticity, the shear strain after the return map should be the same as the elastic strain. Does it mean the elastic strain is very small? Please check this.

Yeah, this may be the issue here.
Here, we calculate the shear strain after return mapping based on the shear stress after return mapping.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/9227ba65-65f0-4949-80f0-9724959063f5)
Initially, we use the way because we thought that, for elasticity, the scale matrix `𝝋` will be the identity matrix; while for plasticity, there is a reduction for the shear stress (due to return mapping and plastic constitutive relation), then the shear strain after return mapping will also decrease. Then we can get a smaller scale matrix `𝝋` for plasticity, i.e., smaller hourglass control for plasticity, which is what we want. 

But, as you mentioned just now, the return mapping only works for shear stress, and theoretically, the shear strain should not change.
I guess the stress-strain relationship is not reversible. We can get the stress from strain, but we can not get strain from stress, because one stress state may corresponds to multiple strains.

DongWuTUM commented:
The shear stress is changing, but the J2 remains the same. I am not sure. Please check it. I think the stress-strain relationship is reversible as they are tensors.

Shuaihao-Zhang commented:
> The shear stress is changing, but the J2 remains the same. I am not sure. Please check it. I think the stress-strain relationship is reversible as they are tensors.

The second invariant J2 also changes after return mapping.
If you check the return mapping process in `I1-J2` (first Invariant and second Invariant ) space, it's clear.
The following figure is the returning mapping in `I1-J2` space for DP constitutive model [1].
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/f26d24a9-7691-4ad1-be0d-61de8b9960e3)

The process B-C is the return mapping for deviatoric stress (I1 keeps unchanged). We can see the J2 (y-axis) also changed. And for J2 plasticity model, the return mapping process is similar with the DP model.

I think the stress-strain relationship is not reversible. We can get the stress from strain as one strain only corresponds to one stress, but we can not get strain from stress because on stress corresponds to several strains, as shown in the figure below.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/4b5070a5-2ae1-46b0-9742-ad3add60149c)

[1]Bui, H. H., Fukagawa, R., Sako, K., Ohno, S., 2008. Lagrangian meshfree particles method (SPH) for large deformation and failure flows of geomaterial using elastic–plastic soil constitutive model. International journal for numerical and analytical methods in geomechanics 32(12), 1537-1570.

DongWuTUM commented:
![summarize](https://github.com/Xiangyu-Hu/SPHinXsys/assets/100197018/4eadc3cb-380d-4aa2-9777-05e4a4465731)

## I have summarized the differences between the TL and UL. 
1. Please check whether it is correct or not.
2. Check the difference in the scaling matrix when running cases of TL and UL.
3. Try: a. include the isotropic term before inverse operation (Please show me how you do it); b. Ɛ^(-1) = (Ɛ+I)^(-1) - I (Here, we can try Ɛ is shear strain or full strain); c. scaling matrix = (Ɛ_corrected+I)(Ɛ+I)^(-1).
4. The TL code is in branch "try_plastic_decomposed". (Note that the factor zeta in the code has not been adjusted. I will upload the newest code tomorrow.) 
5. The shear stress-strain invertibility is not important here as the relation between them is only needed to times 2G.

Xiangyu-Hu commented:
I have the feeling that the inconsistency is coming from the \phi term and \hat{v}_ij term.
Since we are using full strain tensor for the second part of  \hat{V}_ij, the \phi term should also be the same.
The derivation is e_ij -> \phi_ij^-1 * \phi_ij^e_ij -> \phi_ij^-1 * (v_ij - \phi_ij^e_ij) =  \phi_ij^-1 *  \hat{v}_ij.

Xiangyu-Hu commented:
![20231128_094909.jpg](https://github.com/Xiangyu-Hu/SPHinXsys/assets/37415154/99b5479a-9c4d-4f3a-a53c-2950f267e3e8)

The suggested formulation on the correction term. Note that we need compute the effective strain rate within return mapping.

Xiangyu-Hu commented:
![20231128_094909.jpg](https://github.com/Xiangyu-Hu/SPHinXsys/assets/37415154/99b5479a-9c4d-4f3a-a53c-2950f267e3e8)

The suggested formulation on the correction term. Note that we need compute the effective strain rate within return mapping.

Shuaihao-Zhang commented:
@Xiangyu-Hu @DongWuTUM 
Thanks a lot. I will check these in detail.

Shuaihao-Zhang commented:
> ![20231128_094909.jpg](https://private-user-images.githubusercontent.com/37415154/286188425-99b5479a-9c4d-4f3a-a53c-2950f267e3e8.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyMjc1OTIsIm5iZiI6MTcwMTIyNzI5MiwicGF0aCI6Ii8zNzQxNTE1NC8yODYxODg0MjUtOTliNTQ3OWEtOWM0ZC00ZjNhLWE1M2MtMjk1MGYyNjdlM2U4LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDAzMDgxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkxM2Y4NTBmZGQwYTYwNjE3NWUzNTc5NWM4Yjg0MmQzZjg4NWE5OTQyYWE0NjQzOTY2ZjAzOTgwZmE1ODc0NTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.g-kmxh77K1a8NEfg_h7fJ2e_7b_Wj7tDnyXwEgCDwU0)
> 
> The suggested formulation on the correction term. Note that we need compute the effective strain rate within return mapping.

The new form is different from the previous formulation even for elastic case. 
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/8119cd7d-65aa-4155-9944-b732e355ba03)
I tested the new formulation and it doesn't work for elastic oscillating beam. It seems the correction term is too large. I do use `B_ `correction here
![oscillating_beam](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/76f200ef-e68a-421b-878e-ad02f40ff9f4)

Xiangyu-Hu commented:
> > ![20231128_094909.jpg](https://private-user-images.githubusercontent.com/37415154/286188425-99b5479a-9c4d-4f3a-a53c-2950f267e3e8.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyMjc1OTIsIm5iZiI6MTcwMTIyNzI5MiwicGF0aCI6Ii8zNzQxNTE1NC8yODYxODg0MjUtOTliNTQ3OWEtOWM0ZC00ZjNhLWE1M2MtMjk1MGYyNjdlM2U4LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDAzMDgxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkxM2Y4NTBmZGQwYTYwNjE3NWUzNTc5NWM4Yjg0MmQzZjg4NWE5OTQyYWE0NjQzOTY2ZjAzOTgwZmE1ODc0NTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.g-kmxh77K1a8NEfg_h7fJ2e_7b_Wj7tDnyXwEgCDwU0)
> > The suggested formulation on the correction term. Note that we need compute the effective strain rate within return mapping.
> 
> The new form is different from the previous formulation even for elastic case. ![image](https://private-user-images.githubusercontent.com/43446342/286471486-8119cd7d-65aa-4155-9944-b732e355ba03.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyNDAwNTMsIm5iZiI6MTcwMTIzOTc1MywicGF0aCI6Ii80MzQ0NjM0Mi8yODY0NzE0ODYtODExOWNkN2QtNjVhYS00MTU1LTk5NDQtYjczMmUzNTViYTAzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDA2MzU1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY5MDEwYWU3MzM0M2MwYzI5ZTYwYWYxNzhlMWYzMjcyNTE0Y2FhOGIxZTM1NWIzNWIyYzkzNWNkNTdlZDNhMzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9DQd2-i5C1W9pv-3xWUKLQbmnypXmQj32drGa0h7JCA) I tested the new formulation and it doesn't work for elastic oscillating beam. It seems the correction term is too large. I do use `B_ `correction here ![oscillating_beam](https://private-user-images.githubusercontent.com/43446342/286472223-76f200ef-e68a-421b-878e-ad02f40ff9f4.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyNDAwNTMsIm5iZiI6MTcwMTIzOTc1MywicGF0aCI6Ii80MzQ0NjM0Mi8yODY0NzIyMjMtNzZmMjAwZWYtZTY4YS00MjFiLTg3OGUtYWQwMmY0MGZmOWY0LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDA2MzU1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjNWM4ODcyYzJkNjUyZDI0ZWViZjI5MmQwM2NhNWIzYTQyYjI0N2I0MGE5Y2QzNTFkMjIzYWEwOWVmNzJmMmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1bJw8MmCYZnXTXXNiJN9-poMxuIe3wVmgMTIkTzj_S8) [ ![oscillating_beam](https://private-user-images.githubusercontent.com/43446342/286472223-76f200ef-e68a-421b-878e-ad02f40ff9f4.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyNDAwNTMsIm5iZiI6MTcwMTIzOTc1MywicGF0aCI6Ii80MzQ0NjM0Mi8yODY0NzIyMjMtNzZmMjAwZWYtZTY4YS00MjFiLTg3OGUtYWQwMmY0MGZmOWY0LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDA2MzU1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjNWM4ODcyYzJkNjUyZDI0ZWViZjI5MmQwM2NhNWIzYTQyYjI0N2I0MGE5Y2QzNTFkMjIzYWEwOWVmNzJmMmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1bJw8MmCYZnXTXXNiJN9-poMxuIe3wVmgMTIkTzj_S8) ](https://private-user-images.githubusercontent.com/43446342/286472223-76f200ef-e68a-421b-878e-ad02f40ff9f4.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyNDAwNTMsIm5iZiI6MTcwMTIzOTc1MywicGF0aCI6Ii80MzQ0NjM0Mi8yODY0NzIyMjMtNzZmMjAwZWYtZTY4YS00MjFiLTg3OGUtYWQwMmY0MGZmOWY0LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDA2MzU1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjNWM4ODcyYzJkNjUyZDI0ZWViZjI5MmQwM2NhNWIzYTQyYjI0N2I0MGE5Y2QzNTFkMjIzYWEwOWVmNzJmMmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1bJw8MmCYZnXTXXNiJN9-poMxuIe3wVmgMTIkTzj_S8) [ ](https://private-user-images.githubusercontent.com/43446342/286472223-76f200ef-e68a-421b-878e-ad02f40ff9f4.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyNDAwNTMsIm5iZiI6MTcwMTIzOTc1MywicGF0aCI6Ii80MzQ0NjM0Mi8yODY0NzIyMjMtNzZmMjAwZWYtZTY4YS00MjFiLTg3OGUtYWQwMmY0MGZmOWY0LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDA2MzU1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjNWM4ODcyYzJkNjUyZDI0ZWViZjI5MmQwM2NhNWIzYTQyYjI0N2I0MGE5Y2QzNTFkMjIzYWEwOWVmNzJmMmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1bJw8MmCYZnXTXXNiJN9-poMxuIe3wVmgMTIkTzj_S8)

It seems you misunderstood the new formulation, it should recover the elastic exactly before plasticity shows up.

Shuaihao-Zhang commented:
> It seems you misunderstood the new formulation, it should recover the elastic correction exactly before plasticity shows up. Because in that case, the two strain rates, yield identity.

I see. I guess there is a typo in you new formulation.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/b32d3767-49e9-4eeb-b734-d7d14f162728)
The new scale matrix `𝝋` based on the strain rate in ok. But for the ` \hat{v}_ij`, we should still use the old one. The new ` \hat{v}_ij` used the strain rate. The strain rate incudes the transpose of velocity gradient, which we don't have in the old formulation.

Then, I tested the following formulation:
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/c586b2fb-fc6e-46f1-80cf-40afbc077470)
The formulation can reproduce the elastic deformation, But for plasticity, it still crashes like this.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/eec0292b-51ec-4689-9390-3b86355d0853)

Xiangyu-Hu commented:
> > It seems you misunderstood the new formulation, it should recover the elastic correction exactly before plasticity shows up. Because in that case, the two strain rates, yield identity.
> 
> I see. I guess there is a typo in you new formulation. ![image](https://private-user-images.githubusercontent.com/43446342/286523453-b32d3767-49e9-4eeb-b734-d7d14f162728.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyNTAxMTAsIm5iZiI6MTcwMTI0OTgxMCwicGF0aCI6Ii80MzQ0NjM0Mi8yODY1MjM0NTMtYjMyZDM3NjctNDllOS00ZWViLWI3MzQtZDdkMTRmMTYyNzI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDA5MjMzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFjMDc1YWUxN2UzY2RlZDU2YWJlMDlmNzFjYjc5YmNmYzg3YjNkYzRkOTEwOGMwZjNjMzMzMWVmMzUwNmRlZDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.4ZXLhXRkheWrPER3IyLLnPaw_P_Kul6s--6ETP30POc) The new scale matrix `𝝋` based on the strain rate in ok. But for the ` \hat{v}_ij`, we should still use the old one. The new ` \hat{v}_ij` used the strain rate. The strain rate incudes the transpose of velocity gradient, which we don't have in the old formulation.
> 
> Then, I tested the following formulation: ![image](https://private-user-images.githubusercontent.com/43446342/286525489-c586b2fb-fc6e-46f1-80cf-40afbc077470.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyNTAxMTAsIm5iZiI6MTcwMTI0OTgxMCwicGF0aCI6Ii80MzQ0NjM0Mi8yODY1MjU0ODktYzU4NmIyZmItZmM2ZS00NmYxLTgwY2YtNDBhZmJjMDc3NDcwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDA5MjMzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2YWY2OGUxNGYxZGU3M2Q1MWE4MmRhZDBkMjUzZjYwNWMyODUzYzczMGQxOWQ1MjI1NGVlMjE0NmEwODAwOTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.f9VtFFaA3OG7STMlgcyD40ULJVo1CIIjd3wRg4VDqkA) The formulation can reproduce the elastic deformation, But for plasticity, it still crashes like this. ![image](https://private-user-images.githubusercontent.com/43446342/286536191-eec0292b-51ec-4689-9390-3b86355d0853.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyNTAxMTAsIm5iZiI6MTcwMTI0OTgxMCwicGF0aCI6Ii80MzQ0NjM0Mi8yODY1MzYxOTEtZWVjMDI5MmItNTFlYy00Njg5LTkzOTAtM2I4NjM1NWQwODUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDA5MjMzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWEwMDdjNTM3NjM5MmFlMzc1MzFjNDRmYzVmMWFmYTMzYTQ1MzBmZWU2NDZjYzQ4NGMzNWU0MGIyZWM2ZDViMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.VAPJBbWUqmpRjNWdVIAS4guIBX6yGgIJR25F_ZJ9nqw)

Sorry. I missed the notation. I think that we should use inverse of the velocity gradient for the second term, as we are using it for V_hat.

DongWuTUM commented:
@Shuaihao-Zhang I have upload the newest houglass control code to my branch https://github.com/DongWuTUM/SPHinXsys/tree/plastic_hourglass_control. I don't have time to test in Linux, but it already works in Windows.

Xiangyu-Hu commented:
![20231129_121048.jpg](https://github.com/Xiangyu-Hu/SPHinXsys/assets/37415154/07e975fc-996c-49d0-b97a-82971cbeb4d2)

Xiangyu-Hu commented:
> ![20231129_121048.jpg](https://private-user-images.githubusercontent.com/37415154/286574195-07e975fc-996c-49d0-b97a-82971cbeb4d2.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyNTY5NzQsIm5iZiI6MTcwMTI1NjY3NCwicGF0aCI6Ii8zNzQxNTE1NC8yODY1NzQxOTUtMDdlOTc1ZmMtOTk2Yy00OWQwLWI5N2EtODI5NzFjYmViNGQyLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDExMTc1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZjNmEyNzRiMDI4MTg4NjM5ZTgyYTgwNzdkY2YzMTc1ZjdjYjE0OTUzZDBkZWRjMTY0ODQzZGZmMDY1MzkyNGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sYNnG812eiZFCy_rY9W6FGJSiLLmTa40clcz8GOT_kw)

@Shuaihao-Zhang Could you update the code with this formulation? Thanks.

Shuaihao-Zhang commented:
> @Shuaihao-Zhang Could you update the code with this formulation? Thanks.

Sure, I have add the new formulation and the simulation still crashes quick after the bar touches the plate.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/2989a9da-9d42-4eb6-8bd1-ff72728d27dd)
You can check the code at:
https://github.com/Xiangyu-Hu/SPHinXsys/blob/4352844bf976493eee43c7824be28fc2d2a8f7ce/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L551-L561
I used the difference of the shear strain after return mapping between time t and t-1 to calculate the effective shear strain rate.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/c41a2688-3c1b-46ed-bda2-a552f17dafb0)
Then I also added the hydrostatic strain rate to get the effective  total strain rate.
I also tried with only the shear strain rate, and it also crashed.

Shuaihao-Zhang commented:
> @Shuaihao-Zhang I have upload the newest houglass control code to my branch https://github.com/DongWuTUM/SPHinXsys/tree/plastic_hourglass_control. I don't have time to test in Linux, but it already works in Windows.

Thanks! I will check it in detail.

Shuaihao-Zhang commented:
> Please have a check on a simple formulation at the branch https://github.com/Xiangyu-Hu/SPHinXsys/tree/hourglass-control-ul_xy numerical stability seems OK.

Thanks. 
Although the Taylor bar didn't crash with the new formulation, the scale coefficient is not an identity matrix in elastic situations.
https://github.com/Xiangyu-Hu/SPHinXsys/blob/7dacc3d137e6328f1d23ad03e8060f9b65242b03/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L547
If I skip return mapping and use `return shear_stress_rate_elastic;`, ideally, the bar should bounce back like elastic material, but it didn't and the scale matrix is 0 everywhere.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/f9a61012-3f8c-4613-8c15-3d55d4b0ffa4)

Another thing is that I changed my branch to plane stress condition based on your branch for the J2 plasticity.
Previously, I use the plane strain condition, which means I also need to use 3D stress tensor in 2D cases. So I use the `reduceTensor `and `increaseTensor` to transfer from 2D and 3D. But this just affect 2D cases.

I checked Dong's code, and he used plane stress assumption for TLSPH, which means that the stress is 2D for 2D cases and 3D for 3D cases. Since his result is good, I suppose the plans stress is also ok for the J2 plasticity (for soils, it's not ok).
Then I just changed to plane stress condition based on your new branch. I deleted the transform between 2D and 3D tensors to simplify the code and reduce the source of error.

Xiangyu-Hu commented:
> > Please have a check on a simple formulation at the branch https://github.com/Xiangyu-Hu/SPHinXsys/tree/hourglass-control-ul_xy numerical stability seems OK.
> 
> Thanks. Although the Taylor bar didn't crash with the new formulation, the scale coefficient is not an identity matrix in elastic situations.
> 
> https://github.com/Xiangyu-Hu/SPHinXsys/blob/7dacc3d137e6328f1d23ad03e8060f9b65242b03/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L547
> 
> 
> If I skip return mapping and use `return shear_stress_rate_elastic;`, ideally, the bar should bounce back like elastic material, but it didn't and the scale matrix is 0 everywhere.
> ![image](https://private-user-images.githubusercontent.com/43446342/286846824-f9a61012-3f8c-4613-8c15-3d55d4b0ffa4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEzMzMzMDEsIm5iZiI6MTcwMTMzMzAwMSwicGF0aCI6Ii80MzQ0NjM0Mi8yODY4NDY4MjQtZjlhNjEwMTItM2Y4Yy00NjEzLThjMTUtM2Q1NWQ0YjBmZmE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTMwVDA4MzAwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5N2Y2N2MxMDQwNTRjNWRiY2FhMmFkMzg2MTY4ZjMxODJkMTBmY2ZmMGVlM2ZlODhhNzQzZTYwYWI0YTYxYzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.fed-9TnETGpRlzQx6lyogFw-DgFbt7GeKhXQ9ItueKw)
> Another thing is that I changed my branch to plane stress condition based on your branch for the J2 plasticity. Previously, I use the plane strain condition, which means I also need to use 3D stress tensor in 2D cases. So I use the `reduceTensor `and `increaseTensor` to transfer from 2D and 3D. But this just affect 2D cases.
> 
> I checked Dong's code, and he used plane stress assumption for TLSPH, which means that the stress is 2D for 2D cases and 3D for 3D cases. Since his result is good, I suppose the plans stress is also ok for the J2 plasticity (for soils, it's not ok). Then I just changed to plane stress condition based on your new branch. I deleted the transform between 2D and 3D tensors to simplify the code and reduce the source of error.

Xiangyu-Hu commented:
> > > Please have a check on a simple formulation at the branch https://github.com/Xiangyu-Hu/SPHinXsys/tree/hourglass-control-ul_xy numerical stability seems OK.
> > 
> > 
> > Thanks. Although the Taylor bar didn't crash with the new formulation, the scale coefficient is not an identity matrix in elastic situations.
> > https://github.com/Xiangyu-Hu/SPHinXsys/blob/7dacc3d137e6328f1d23ad03e8060f9b65242b03/tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_inner.cpp#L547
> > 
> > If I skip return mapping and use `return shear_stress_rate_elastic;`, ideally, the bar should bounce back like elastic material, but it didn't and the scale matrix is 0 everywhere.
> > ![image](https://private-user-images.githubusercontent.com/43446342/286846824-f9a61012-3f8c-4613-8c15-3d55d4b0ffa4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEzMzMzMDEsIm5iZiI6MTcwMTMzMzAwMSwicGF0aCI6Ii80MzQ0NjM0Mi8yODY4NDY4MjQtZjlhNjEwMTItM2Y4Yy00NjEzLThjMTUtM2Q1NWQ0YjBmZmE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTMwVDA4MzAwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5N2Y2N2MxMDQwNTRjNWRiY2FhMmFkMzg2MTY4ZjMxODJkMTBmY2ZmMGVlM2ZlODhhNzQzZTYwYWI0YTYxYzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.fed-9TnETGpRlzQx6lyogFw-DgFbt7GeKhXQ9ItueKw)
> > Another thing is that I changed my branch to plane stress condition based on your branch for the J2 plasticity. Previously, I use the plane strain condition, which means I also need to use 3D stress tensor in 2D cases. So I use the `reduceTensor `and `increaseTensor` to transfer from 2D and 3D. But this just affect 2D cases.
> > I checked Dong's code, and he used plane stress assumption for TLSPH, which means that the stress is 2D for 2D cases and 3D for 3D cases. Since his result is good, I suppose the plans stress is also ok for the J2 plasticity (for soils, it's not ok). Then I just changed to plane stress condition based on your new branch. I deleted the transform between 2D and 3D tensors to simplify the code and reduce the source of error.

It works for elastic (both elastic shear stress rate and no returnning map) 
3D case as shown in the test branch https://github.com/Xiangyu-Hu/SPHinXsys/tree/hourglass-control-ul_xy
Note that the correction is numerical force term, there is no requirement for the consistency mentioned by you. More important is they approaching zero when resolution is high and it is small value compared to the main stress induced force.
I made small change on the scaling factor, however i do not known which is better.

Shuaihao-Zhang commented:
> 
> It works for elastic (both elastic shear stress rate and no returnning map) 3D case as shown in the test branch https://github.com/Xiangyu-Hu/SPHinXsys/tree/hourglass-control-ul_xy Note that the correction is numerical force term, there is no requirement for the consistency mentioned by you. More important is they approaching zero when resolution is high and it is small value compared to the main stress induced force. I made small change on the scaling factor, however i do not known which is better.

Thanks. I checked you branch. The 3D Taylor bar seems elastic if we use elastic shear stress rate and no return mapping.
But the scale matrix if super large, and theoretically, it should be an identity matrix for elasticity.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/99c0bfbe-9881-490b-8996-064cb3d06bd1)
Also for 2D Taylor bar, the scale matrix is very large although is didn't crash.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/dc94324f-6ef6-47e1-89a7-eb5f4a1cb829)
For 2D oscillating beam, if I use the plastic constitutive relation (with elastic shear stress rate and no return mapping), the result should be exactly the same with the elastic beam (for previous method, we can total reproduce this). But it crashes and I think this is related to the unreasonable scale matrix.
![image](https://github.com/Xiangyu-Hu/SPHinXsys/assets/43446342/2975ef21-dc45-4243-b4b9-aa3f42ad4a6d)

Shuaihao-Zhang commented:
> @Shuaihao-Zhang It is the plane strain (not plane stress) problem for 2D case. The stress in third coordinate is not needed for calculation. (I remember the stress in third coordinate is needed for soil. You can ignore this stress now.)

Thanks! Then I will just use 2D matrix for 2D cases.

Xiangyu-Hu commented:
@Shuaihao-Zhang  @DongWuTUM 
Finally, matrix inverse regularization is done. Now oscillation beam and Taylor bar both work.
[shear_stress_try ](https://github.com/Xiangyu-Hu/SPHinXsys/tree/ul_plastic_hourglass_try)

Shuaihao-Zhang commented:
> @Shuaihao-Zhang @DongWuTUM Finally, matrix inverse regularization is done. Now oscillation beam and Taylor bar both work. [shear_stress_try ](https://github.com/Xiangyu-Hu/SPHinXsys/tree/ul_plastic_hourglass_try)

Thanks! Now the elastic deformation can be totally reproduced.
